crates/prover/src/runner.rs

```diff
@@ -202,85 +299,139 @@ where
-            })
-            .await?
-            .height();
+        // Save the genesis state root to ledger db
+        if self.start_l2_height == 1 {
+            self.ledger_db
+                .set_l2_state_root(0, &self.state_root.clone())?;
+        }
 
-            if l1_height > last_finalized_height {
-                sleep(Duration::from_secs(1)).await;
-                continue;
+        let (l1_tx, mut l1_rx) = mpsc::channel(1);
+        let da_service = self.da_service.clone();
+        let l1_block_cache = self.l1_block_cache.clone();
+        let l1_handle = tokio::spawn(async move {
+            l1_sync(start_l1_height, da_service, l1_tx, l1_block_cache).await;
+        });
+        tokio::pin!(l1_handle);
+
+        let (l2_tx, mut l2_rx) = mpsc::channel(1);
+        let start_l2_height = self.start_l2_height;
+        let sequencer_client = self.sequencer_client.clone();
+        let sync_blocks_count = self.sync_blocks_count;
+        let l2_handle = tokio::spawn(async move {
+            sync_l2::<Da>(start_l2_height, sequencer_client, l2_tx, sync_blocks_count).await;
+        });
+        tokio::pin!(l2_handle);
+        let da_service = self.da_service.clone();
+        let l1_block_cache = self.l1_block_cache.clone();
+        let mut pending_l1_blocks: VecDeque<<Da as DaService>::FilteredBlock> =
+            VecDeque::<Da::FilteredBlock>::new();
+        let pending_l1 = &mut pending_l1_blocks;
+        loop {
+            select! {
+                _ = &mut l1_handle => {panic!("l1 sync handle exited unexpectedly");},
+                _ = &mut l2_handle => {panic!("l2 sync handle exited unexpectedly");},
+                Some(l1_block) = l1_rx.recv() => {
+                    pending_l1.push_back(l1_block);
+                 },
+                _ = sleep(Duration::from_secs(1)) => {
+                    self.run_inner(
+                        pending_l1,
+                        skip_submission_until_l1,
+                        &pg_client, &prover_config,
+                    ).await;
+                },
+                Some(l2_blocks) = l2_rx.recv() => {
+                    for (l2_height, l2_block) in l2_blocks {
+                        let l1_block = get_da_block_at_height(&da_service, l2_block.da_slot_height, l1_block_cache.clone()).await?;
+                        if let Err(e) = self.process_l2_block(l2_height, l2_block, l1_block).await {
+                            error!("Could not process L2 block: {}", e);
+                        }
+                    }
+                },
             }
+        }
+    }
 
-            let filtered_block = retry_backoff(exponential_backoff.clone(), || async {
-                da_service
-                    .get_block_at(l1_height)
-                    .await
-                    .map_err(backoff::Error::transient)
-            })
-            .await?;
-
-            // map the height to the hash
+    async fn run_inner(
```

Comment:
I would have hoped that this would be broken down into different components:
1. X would be executed only when an L2 is received
2. Y would be executed only when an L1 is received
3. Z would wait for both L1/L2

This is basically the meat of the prover,,, IMO we did split L1/L2 sync in this PR but we didn't split this.

Reply:
could you elaborate please?

Reply:
This method:
1. Processes pending L1 blocks
2. Processes sequencer commitments
3. Applies soft confirmations

- It can be broken down into multiple methods
- Applying soft confirmation can happen when we receive an L2
- Processing commitments can happen when we receive an L1
So calling the sub methods can be called in their more appropriate places.